File: learning-factory/warehouse/artifacts/projects/build-your-own-x/remediation/27debb2830b4885fa831/repair-v1/job_byox_repair_v1_g1_f7de7abfdff812ca5a1921818f925788/attempt-001/sealed/reference/src/parser.js
import { ParseError, boundedInteger } from "./errors.js";
import { TokenType as T } from "./tokens.js";

const MAX_PARSE_DEPTH = 1_000;

export function parse(tokens, options = {}) {
  if (!Array.isArray(tokens)) throw new TypeError("tokens must be an array");
  const maxParseDepth = boundedInteger(options, "maxParseDepth", MAX_PARSE_DEPTH);
  return new Parser(tokens, maxParseDepth).parse();
}

class Parser {
  constructor(tokens, maxDepth) {
    this.tokens = tokens;
    this.maxDepth = maxDepth;
    this.current = 0;
    this.depth = 0;
    if (tokens.length === 0 || tokens[tokens.length - 1]?.type !== T.EOF) {
      throw new ParseError("Token stream must end with EOF", tokenLocation(tokens.at(-1)));
    }
  }

  parse() {
    const start = tokenLocation(this._peek());
    const body = [];
    while (!this._check(T.EOF)) body.push(this._statementGuarded());
    this._consume(T.EOF, "Expected end of input");
    if (this.current !== this.tokens.length) {
      throw new ParseError("Tokens found after EOF", tokenLocation(this.tokens[this.current]));
    }
    return { type: "Program", body, loc: start };
  }

  _statementGuarded() { return this._guard(() => this._statement()); }

  _statement() {
    if (this._match(T.LET)) return this._letStatement(this._previous());
    if (this._match(T.PRINT)) return this._printStatement(this._previous());
    if (this._match(T.IF)) return this._ifStatement(this._previous());
    if (this._match(T.WHILE)) return this._whileStatement(this._previous());
    if (this._match(T.LEFT_BRACE)) return this._block(this._previous());
    return this._expressionStatement();
  }

  _letStatement(keyword) {
    const nameToken = this._consume(T.IDENTIFIER, "Expected variable name after 'let'");
    this._consume(T.EQUAL, "Expected '=' after variable name");
    const initializer = this._expression();
    this._consume(T.SEMICOLON, "Expected ';' after declaration");
    return {
      type: "LetStatement",
      name: identifier(nameToken),
      initializer,
      loc: tokenLocation(keyword)
    };
  }

  _printStatement(keyword) {
    const expression = this._expression();
    this._consume(T.SEMICOLON, "Expected ';' after printed expression");
    return { type: "PrintStatement", expression, loc: tokenLocation(keyword) };
  }

  _expressionStatement() {
    const expression = this._expression();
    this._consume(T.SEMICOLON, "Expected ';' after expression");
    return { type: "ExpressionStatement", expression, loc: expression.loc };
  }

  _ifStatement(keyword) {
    this._consume(T.LEFT_PAREN, "Expected '(' after 'if'");
    const test = this._expression();
    this._consume(T.RIGHT_PAREN, "Expected ')' after condition");
    const open = this._consume(T.LEFT_BRACE, "Expected block after if condition");
    const consequent = this._block(open);
    let alternate = null;
    if (this._match(T.ELSE)) {
      const elseOpen = this._consume(T.LEFT_BRACE, "Expected block after 'else'");
      alternate = this._block(elseOpen);
    }
    return { type: "IfStatement", test, consequent, alternate, loc: tokenLocation(keyword) };
  }

  _whileStatement(keyword) {
    this._consume(T.LEFT_PAREN, "Expected '(' after 'while'");
    const test = this._expression();
    this._consume(T.RIGHT_PAREN, "Expected ')' after condition");
    const open = this._consume(T.LEFT_BRACE, "Expected block after while condition");
    return { type: "WhileStatement", test, body: this._block(open), loc: tokenLocation(keyword) };
  }

  _block(open) {
    const body = [];
    while (!this._check(T.RIGHT_BRACE) && !this._check(T.EOF)) body.push(this._statementGuarded());
    this._consume(T.RIGHT_BRACE, "Expected '}' after block");
    return { type: "BlockStatement", body, loc: tokenLocation(open) };
  }

  _expression() { return this._guard(() => this._assignment()); }

  _assignment() {
    const left = this._or();
    if (!this._match(T.EQUAL)) return left;
    const equals = this._previous();
    const value = this._guard(() => this._assignment());
    if (left.type !== "Identifier") {
      throw new ParseError("Invalid assignment target", tokenLocation(equals));
    }
    return { type: "AssignmentExpression", name: left, value, loc: left.loc };
  }

  _or() {
    let expression = this._and();
    while (this._match(T.OR)) {
      const operator = this._previous();
      expression = binaryNode("LogicalExpression", expression, operator, this._and());
    }
    return expression;
  }

  _and() {
    let expression = this._equality();
    while (this._match(T.AND)) {
      const operator = this._previous();
      expression = binaryNode("LogicalExpression", expression, operator, this._equality());
    }
    return expression;
  }

  _equality() {
    let expression = this._comparison();
    while (this._match(T.EQUAL_EQUAL, T.BANG_EQUAL)) {
      const operator = this._previous();
      expression = binaryNode("BinaryExpression", expression, operator, this._comparison());
    }
    return expression;
  }

  _comparison() {
    let expression = this._term();
    while (this._match(T.GREATER, T.GREATER_EQUAL, T.LESS, T.LESS_EQUAL)) {
      const operator = this._previous();
      expression = binaryNode("BinaryExpression", expression, operator, this._term());
    }
    return expression;
  }

  _term() {
    let expression = this._factor();
    while (this._match(T.PLUS, T.MINUS)) {
      const operator = this._previous();
      expression = binaryNode("BinaryExpression", expression, operator, this._factor());
    }
    return expression;
  }

  _factor() {
    let expression = this._unary();
    while (this._match(T.STAR, T.SLASH)) {
      const operator = this._previous();
      expression = binaryNode("BinaryExpression", expression, operator, this._unary());
    }
    return expression;
  }

  _unary() {
    if (this._match(T.BANG, T.MINUS)) {
      const operator = this._previous();
      return {
        type: "UnaryExpression",
        operator: operator.lexeme,
        argument: this._guard(() => this._unary()),
        loc: tokenLocation(operator)
      };
    }
    return this._primary();
  }

  _primary() {
    if (this._match(T.FALSE)) return literal(false, this._previous());
    if (this._match(T.TRUE)) return literal(true, this._previous());
    if (this._match(T.NULL)) return literal(null, this._previous());
    if (this._match(T.NUMBER, T.STRING)) return literal(this._previous().literal, this._previous());
    if (this._match(T.IDENTIFIER)) return identifier(this._previous());
    if (this._match(T.LEFT_PAREN)) {
      const expression = this._expression();
      this._consume(T.RIGHT_PAREN, "Expected ')' after expression");
      return expression;
    }
    throw new ParseError("Expected expression", tokenLocation(this._peek()));
  }

  _guard(callback) {
    this.depth += 1;
    if (this.depth > this.maxDepth) {
      this.depth -= 1;
      throw new ParseError(`Parse depth exceeds ${this.maxDepth}`, tokenLocation(this._peek()));
    }
    try { return callback(); } finally { this.depth -= 1; }
  }

  _match(...types) {
    for (const type of types) {
      if (this._check(type)) { this._advance(); return true; }
    }
    return false;
  }

  _consume(type, message) {
    if (this._check(type)) return this._advance();
    throw new ParseError(message, tokenLocation(this._peek()));
  }

  _check(type) { return this._peek()?.type === type; }
  _advance() { if (this.current < this.tokens.length) this.current += 1; return this._previous(); }
  _peek() { return this.tokens[this.current]; }
  _previous() { return this.tokens[this.current - 1]; }
}

function tokenLocation(token) {
  return {
    line: Number.isInteger(token?.line) ? token.line : null,
    column: Number.isInteger(token?.column) ? token.column : null
  };
}
function identifier(token) { return { type: "Identifier", name: token.lexeme, loc: tokenLocation(token) }; }
function literal(value, token) { return { type: "Literal", value, loc: tokenLocation(token) }; }
function binaryNode(type, left, operator, right) {
  return { type, operator: operator.lexeme, left, right, loc: tokenLocation(operator) };
}

File: learning-factory/warehouse/artifacts/projects/build-your-own-x/remediation/40bc68c01215515368f9/baseline-b561a5be16bf51e90720/repair-v1/job_byox_repair_s2_v1_g1_ec7f6cd87bd8b8d0c42fc38d0df807d4/attempt-001/sealed/reference/src/main/java/org/learningfactory/mica/Parser.java
package org.learningfactory.mica;

import java.util.ArrayList;
import java.util.List;

public final class Parser {
    private final List<Token> tokens;
    private int current;
    private List<Stmt> parsed;

    public Parser(List<Token> tokens) {
        if (tokens == null || tokens.isEmpty()) {
            throw new MicaException(MicaException.Kind.PARSE, 1, 1, "token stream must contain EOF");
        }
        this.tokens = List.copyOf(tokens);
    }

    public List<Stmt> parse() {
        if (parsed != null) return parsed;
        List<Stmt> statements = new ArrayList<>();
        while (!isAtEnd()) statements.add(declaration());
        parsed = List.copyOf(statements);
        return parsed;
    }

    private Stmt declaration() {
        if (match(TokenType.LET)) return letDeclaration();
        return statement();
    }

    private Stmt letDeclaration() {
        Token name = consume(TokenType.IDENTIFIER, "expected variable name after 'let'");
        consume(TokenType.EQUAL, "expected '=' after variable name");
        Expr initializer = expression();
        consume(TokenType.SEMICOLON, "expected ';' after variable declaration");
        return new Stmt.Let(name, initializer);
    }

    private Stmt statement() {
        if (match(TokenType.PRINT)) return printStatement(previous());
        if (match(TokenType.IF)) return ifStatement(previous());
        if (match(TokenType.WHILE)) return whileStatement(previous());
        if (match(TokenType.LEFT_BRACE)) return new Stmt.Block(previous(), block());
        return expressionStatement();
    }

    private Stmt printStatement(Token keyword) {
        Expr value = expression();
        consume(TokenType.SEMICOLON, "expected ';' after value");
        return new Stmt.Print(keyword, value);
    }

    private Stmt ifStatement(Token keyword) {
        consume(TokenType.LEFT_PAREN, "expected '(' after 'if'");
        Expr condition = expression();
        consume(TokenType.RIGHT_PAREN, "expected ')' after condition");
        Stmt thenBranch = statement();
        Stmt elseBranch = null;
        if (match(TokenType.ELSE)) elseBranch = statement();
        return new Stmt.If(keyword, condition, thenBranch, elseBranch);
    }

    private Stmt whileStatement(Token keyword) {
        consume(TokenType.LEFT_PAREN, "expected '(' after 'while'");
        Expr condition = expression();
        consume(TokenType.RIGHT_PAREN, "expected ')' after condition");
        return new Stmt.While(keyword, condition, statement());
    }

    private List<Stmt> block() {
        List<Stmt> statements = new ArrayList<>();
        while (!check(TokenType.RIGHT_BRACE) && !isAtEnd()) statements.add(declaration());
        consume(TokenType.RIGHT_BRACE, "expected '}' after block");
        return statements;
    }

    private Stmt expressionStatement() {
        Expr value = expression();
        Token terminator = consume(TokenType.SEMICOLON, "expected ';' after expression");
        return new Stmt.Expression(value, terminator);
    }

    private Expr expression() {
        return assignment();
    }

    private Expr assignment() {
        Expr expression = or();
        if (match(TokenType.EQUAL)) {
            Token equals = previous();
            Expr value = assignment();
            if (expression instanceof Expr.Variable variable) return new Expr.Assign(variable.name(), value);
            throw error(equals, "invalid assignment target");
        }
        return expression;
    }

    private Expr or() {
        Expr expression = and();
        while (match(TokenType.OR)) expression = new Expr.Logical(expression, previous(), and());
        return expression;
    }

    private Expr and() {
        Expr expression = equality();
        while (match(TokenType.AND)) expression = new Expr.Logical(expression, previous(), equality());
        return expression;
    }

    private Expr equality() {
        Expr expression = comparison();
        while (match(TokenType.BANG_EQUAL, TokenType.EQUAL_EQUAL)) {
            expression = new Expr.Binary(expression, previous(), comparison());
        }
        return expression;
    }

    private Expr comparison() {
        Expr expression = term();
        while (match(TokenType.GREATER, TokenType.GREATER_EQUAL, TokenType.LESS, TokenType.LESS_EQUAL)) {
            expression = new Expr.Binary(expression, previous(), term());
        }
        return expression;
    }

    private Expr term() {
        Expr expression = factor();
        while (match(TokenType.MINUS, TokenType.PLUS)) {
            expression = new Expr.Binary(expression, previous(), factor());
        }
        return expression;
    }

    private Expr factor() {
        Expr expression = unary();
        while (match(TokenType.SLASH, TokenType.STAR)) {
            expression = new Expr.Binary(expression, previous(), unary());
        }
        return expression;
    }

    private Expr unary() {
        if (match(TokenType.BANG, TokenType.MINUS)) return new Expr.Unary(previous(), unary());
        return primary();
    }

    private Expr primary() {
        if (match(TokenType.FALSE)) return new Expr.Literal(previous(), false);
        if (match(TokenType.TRUE)) return new Expr.Literal(previous(), true);
        if (match(TokenType.NIL)) return new Expr.Literal(previous(), null);
        if (match(TokenType.NUMBER, TokenType.STRING)) return new Expr.Literal(previous(), previous().literal());
        if (match(TokenType.IDENTIFIER)) return new Expr.Variable(previous());
        if (match(TokenType.LEFT_PAREN)) {
            Token open = previous();
            Expr expression = expression();
            consume(TokenType.RIGHT_PAREN, "expected ')' after expression");
            return new Expr.Grouping(open, expression);
        }
        throw error(peek(), "expected expression");
    }

    private boolean match(TokenType... types) {
        for (TokenType type : types) {
            if (check(type)) {
                advance();
                return true;
            }
        }
        return false;
    }

    private Token consume(TokenType type, String message) {
        if (check(type)) return advance();
        throw error(peek(), message);
    }

    private boolean check(TokenType type) {
        if (current >= tokens.size()) return false;
        return peek().type() == type;
    }

    private Token advance() {
        if (current < tokens.size()) current++;
        return previous();
    }

    private boolean isAtEnd() {
        return check(TokenType.EOF);
    }

    private Token peek() {
        if (current < tokens.size()) return tokens.get(current);
        Token last = tokens.get(tokens.size() - 1);
        throw error(last, "token stream is missing EOF");
    }

    private Token previous() {
        return tokens.get(current - 1);
    }

    private static MicaException error(Token token, String message) {
        return new MicaException(MicaException.Kind.PARSE, token.line(), token.column(), message);
    }
}
